network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs

```diff
@@ -0,0 +1,333 @@
+import           Network.Mux.Types
+
+
+-- | Message frames passed through the network.
+--
+data Message =
+      MsgClose
+    | MsgBytes BL.ByteString
+  deriving Eq
+
+
+--
+-- QueueChannel
+--
+
+-- | 'QueueChannel' is the low level bearer used by the simulated snocket.
+--
+-- Each read / write queues can be closed independently.  Read queue is closed
+-- once 'MsgClose' is read from the queue; dually, write queue is closed once
+-- 'MsgClose' is written.
+--
+data QueueChannel m = QueueChannel {
+    qcRead        :: StrictTVar m (Maybe (TQueue m Message)),
+    qcWrite       :: StrictTVar m (Maybe (TQueue m Message))
+  }
+
+--
+-- QueueChannel API
+--
+
+readQueueChannel :: ( MonadSTM        m
+                    , MonadThrow (STM m)
+                    )
+                 => QueueChannel m -> m Message
+readQueueChannel QueueChannel { qcRead } =
+    atomically $ do
+      a <- readTVar qcRead >>= traverse readTQueue
+      case a of
+        Nothing           -> throwSTM (resourceVanishedIOError "sim.readQueueChannel")
+        Just msg@MsgClose -> writeTVar qcRead Nothing
+                          >> return msg
+        Just msg          -> return msg
+
+
+writeQueueChannel :: MonadSTM m
+                  => QueueChannel m -> Message -> m Bool
+writeQueueChannel QueueChannel { qcWrite } msg =
+    atomically $ do
+      mq <- readTVar qcWrite
+      case mq of
+        Nothing -> return False
+        Just q  -> writeTQueue q msg
+                >> case msg of
+                      MsgClose -> writeTVar qcWrite Nothing
+                      _        -> return ()
+                >> return True
+
+
+newConnectedQueueChannelPair :: MonadSTM m
+                             => STM m ( QueueChannel m
+                                      , QueueChannel m )
+newConnectedQueueChannelPair = do
+    read  <- newTQueue
+    write <- newTQueue
+    q  <- QueueChannel <$> newTVar (Just read)
+                       <*> newTVar (Just write)
+    q' <- QueueChannel <$> newTVar (Just write)
+                       <*> newTVar (Just read)
+    return (q, q')
+
+
+--
+-- AttenuatedChannel
+--
+
+
+-- | An AttenuatedChannel supports:
+--
+-- - attenuation applied after reading a message from 'QueueChannel';
+-- - two-way close handshake with 120s timeout.  Read side is closed as soon as
+--   an internal 'MsgClose' is received, write side has to be closed explicietly.
+--
+data AttenuatedChannel m = AttenuatedChannel {
+    acRead  :: m BL.ByteString,
+    acWrite :: BL.ByteString -> m (),
+    acClose :: m ()
+  }
+
+
+-- | Error thrown when expected 'MsgClose' does not arrive within `120s`.
+--
+-- This excpetion should not be handled by a simulation, it is a proof of a half
+-- closed connection, which should never happen when using connection manager.
+--
+data CloseError = CloseTimeoutError
+  deriving (Show, Typeable)
+
+instance Exception CloseError
+
+
+data SuccessOrFailure = Success | Failure
+
+type Size = Int64
```

Comment:
Why is this `Int64` and not `Word64`?

Reply:
The only reason is that lazy bytestring is using `Int64` for [`length`](https://hackage.haskell.org/package/bytestring-0.11.1.0/docs/Data-ByteString-Lazy.html#v:length).